tests/frontend/views/test_app.py

```diff
@@ -0,0 +1,224 @@
+    return app
+
+
+@pytest.yield_fixture(autouse=True)
+def repos(github):
+    test_repo = {
+        'full_name': 'Test App',
+        'html_url': 'https://github.com/moj-analytical-services/test_app',
+    }
+    org = github.get_organization.return_value
+    org.get_repos.return_value = [test_repo]
+    github.get_repo.return_value = test_repo
+    yield github
+
+
+@pytest.fixture(autouse=True)
+def s3buckets(app):
+    buckets = {
+        'not_connected': mommy.make('api.S3Bucket'),
+        'connected': mommy.make('api.S3Bucket'),
+    }
+    return buckets
+
+
+@pytest.fixture
+def apps3bucket(app, s3buckets):
+    return mommy.make('api.AppS3Bucket', app=app, s3bucket=s3buckets['connected'])
+
+
+def list(client, *args):
+    return client.get(reverse('list-apps'))
+
+
+def list_all(client, *args):
+    return client.get(reverse('list-all-apps'))
+
+
+def detail(client, app, *args):
+    return client.get(reverse('manage-app', kwargs={'pk': app.id}))
+
+
+def create(client, *args):
+    data = {
+        'repo_url': 'https://github.com/moj-analytical-services/test_app',
+        'connect_bucket': 'later',
+    }
+    return client.post(reverse('create-app'), data)
+
+
+def delete(client, app, *args):
+    return client.post(reverse('delete-app', kwargs={'pk': app.id}))
+
+
+def add_admin(client, app, users, *args):
+    data = {
+        'user_id': users['other_user'].auth0_id,
+    }
+    return client.post(reverse('add-app-admin', kwargs={'pk': app.id}), data)
+
+
+def revoke_admin(client, app, users, *args):
+    kwargs = {
+        'pk': app.id,
+        'user_id': users['app_admin'].auth0_id,
+    }
+    return client.post(reverse('revoke-app-admin', kwargs=kwargs))
+
+
+def add_customers(client, app, *args):
+    data = {
+        'customer_email': 'test@example.com',
+    }
+    return client.post(reverse('add-app-customers', kwargs={'pk': app.id}), data)
+
+
+def remove_customers(client, app, *args):
+    data = {
+        'customer': 'email|user_1',
+    }
+    return client.post(reverse('remove-app-customer', kwargs={'pk': app.id}), data)
+
+
+def connect_bucket(client, app, _, s3buckets, *args):
+    data = {
+        'datasource': s3buckets['not_connected'].id,
+        'access_level': 'readonly',
+    }
+    return client.post(reverse('grant-app-access', kwargs={'pk': app.id}), data)
+
+
+@pytest.mark.parametrize(
+    'view,user,expected_status',
+    [
+        (list, 'superuser', status.HTTP_200_OK),
+        (list, 'app_admin', status.HTTP_200_OK),
+        (list, 'normal_user', status.HTTP_200_OK),
+
+        (list_all, 'superuser', status.HTTP_200_OK),
+        (list_all, 'app_admin', status.HTTP_403_FORBIDDEN),
+        (list_all, 'normal_user', status.HTTP_403_FORBIDDEN),
+
+        (detail, 'superuser', status.HTTP_200_OK),
+        (detail, 'app_admin', status.HTTP_200_OK),
+        (detail, 'normal_user', status.HTTP_403_FORBIDDEN),
+
+        (create, 'superuser', status.HTTP_302_FOUND),
```

Comment:
`302 FOUND`, weird...a redirect? 

[MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/302) seems to suggest it should only be used for `GET` or `HEAD` 🤷‍♂ :

> Even if the specification requires the method (and the body) not to be altered when the redirection is performed, not all user-agents conform here [...]. It is therefore recommended to set the 302 code only as a response for GET or HEAD methods

Also, the description says:

> `302 Found` redirect status response code indicates that the resource requested has been temporarily moved [...]

Maybe plain old `201 Created`/`204 No Content` would be more appropriate for this stuff, but I only had a quick look.

Reply:
These are frontend views, not a REST API. 302 is the status code that Django uses for `RedirectView` class-based views. Arguably, it should be 303.